src/app/server/AclStorage.cpp

```diff
@@ -0,0 +1,488 @@
+    case StagingPrivilege::kManage:
+        to = Privilege::kManage;
+        break;
+    case StagingPrivilege::kAdminister:
+        to = Privilege::kAdminister;
+        break;
+    default:
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Convert(NodeId from, StagingSubject & to)
+{
+    if (IsOperationalNodeId(from) || IsCASEAuthTag(from))
+    {
+        to = { .nodeId = from, .authMode = StagingAuthMode::kCase };
+    }
+    else if (IsGroupId(from))
+    {
+        to = { .nodeId = GroupIdFromNodeId(from), .authMode = StagingAuthMode::kGroup };
+    }
+    else if (IsPAKEKeyId(from))
+    {
+        to = { .nodeId = PAKEKeyIdFromNodeId(from), .authMode = StagingAuthMode::kPase };
+    }
+    else
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Convert(StagingSubject from, NodeId & to)
+{
+    switch (from.authMode)
+    {
+    case StagingAuthMode::kPase:
+        ReturnErrorCodeIf((from.nodeId & ~kMaskPAKEKeyId) != 0, CHIP_ERROR_INVALID_ARGUMENT);
+        to = NodeIdFromPAKEKeyId(static_cast<PasscodeId>(from.nodeId));
+        break;
+    case StagingAuthMode::kCase:
+        to = from.nodeId;
+        break;
+    case StagingAuthMode::kGroup:
+        ReturnErrorCodeIf((from.nodeId & ~kMaskGroupId) != 0, CHIP_ERROR_INVALID_ARGUMENT);
+        to = NodeIdFromGroupId(static_cast<GroupId>(from.nodeId));
+        break;
+    default:
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Convert(const Target & from, StagingTarget & to)
+{
+    if ((from.flags & Target::kCluster) != 0)
+    {
+        to.cluster.SetNonNull(from.cluster);
+    }
+    else
+    {
+        to.cluster.SetNull();
+    }
+    if ((from.flags & Target::kEndpoint) != 0)
+    {
+        to.endpoint.SetNonNull(from.endpoint);
+    }
+    else
+    {
+        to.endpoint.SetNull();
+    }
+    if ((from.flags & Target::kDeviceType) != 0)
+    {
+        to.deviceType.SetNonNull(from.deviceType);
+    }
+    else
+    {
+        to.deviceType.SetNull();
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Convert(const StagingTarget & from, Target & to)
+{
+    to.flags = 0;
+    if (!from.cluster.IsNull())
+    {
+        to.flags |= Target::kCluster;
+        to.cluster = from.cluster.Value();
+    }
+    if (!from.endpoint.IsNull())
+    {
+        to.flags |= Target::kEndpoint;
+        to.endpoint = from.endpoint.Value();
+    }
+    if (!from.deviceType.IsNull())
+    {
+        to.flags |= Target::kDeviceType;
+        to.deviceType = from.deviceType.Value();
+    }
+    return CHIP_NO_ERROR;
+}
+
+class : public EntryListener
+{
+public:
+    void OnEntryChanged(const SubjectDescriptor * subjectDescriptor, FabricIndex fabric, size_t index, const Entry * entry,
+                        ChangeType changeType) override
+    {
+        CHIP_ERROR err;
+
+        auto & storage = Server::GetInstance().GetPersistentStorage();
```

Comment:
Do not reach into Server for the PersistentStorage. We inject the AccessControl::Delegate to the Server *SPECIFICALLY* so that the init of it injects the persistent storage delegate. It has to be kept this way.

Reply:
This is from the other PR. At any rate, that adjustment can be a followup.

Reply:
I will do this on the other PR.